### PR TITLE
Node driver

### PR DESCRIPTION
**Pull request by aiwantaozi**
Related Issue:  
https://github.com/rancher/rancher/issues/12269
Generate driver name for different format driver URL
https://github.com/rancher/rancher/issues/11039
Add driver executable check.

https://github.com/rancher/types/pull/319

### CODE REVIEW

**Comment by aiwantaozi** [comment]
#12269 
#11039

File: pkg/controllers/management/nodedriver/dynamic_driver.go [hunk 1]
```diff
@@ -236,10 +236,20 @@ func (d *Driver) copyBinary(cacheFile, input string) (string, error) {
...
+
+		s := strings.TrimPrefix(path.Base(u.Path), "docker-machine-driver-")
+		name := strings.FieldsFunc(s, func(r rune) bool {
+			return r == '-' || r == '_' || r == '.'
+		})[0]
+
+		if name == "" {
+			return "", fmt.Errorf("invalid URL %s, NAME is empty, path should be of the format docker-machine-driver-NAME", d.url)
+		}
+		driverName = "docker-machine-driver-" + name
```
Comment (StrongMonkey) [inline pkg/controllers/management/nodedriver/dynamic_driver.go]: should be driverName=name?

Comment (StrongMonkey) [reply pkg/controllers/management/nodedriver/dynamic_driver.go]: nvm, can have the prefix

File: pkg/controllers/management/nodedriver/machine_driver.go [hunk 2]
```diff
@@ -71,19 +71,30 @@ func (m *Lifecycle) download(obj *v3.NodeDriver) (*v3.NodeDriver, error) {
...
+	})
+	if err != nil {
+		return obj, err
+	}
 
+	obj = newObj.(*v3.NodeDriver)
+	newObj, err = v3.NodeDriverConditionInstalled.Once(obj, func() (runtime.Object, error) {
 		if err := driver.Install(); err != nil {
-			logrus.Errorf("Failed to download/install driver %s: %v", driver.Name(), err)
+			logrus.Errorf("Failed to install driver %s: %v", driver.Name(), err)
```
Comment (StrongMonkey) [inline pkg/controllers/management/nodedriver/machine_driver.go]: Don't log error in here, error should be thrown to the framework

File: pkg/controllers/management/nodedriver/machine_driver.go [hunk 3]
```diff
@@ -71,19 +71,30 @@ func (m *Lifecycle) download(obj *v3.NodeDriver) (*v3.NodeDriver, error) {
...
+			return nil, err
+		}
+		if err = driver.Excutable(); err != nil {
+			logrus.Errorf("Failed to execute driver %s: %v", driver.Name(), err)
```
Comment (StrongMonkey) [inline pkg/controllers/management/nodedriver/machine_driver.go]: same above

File: pkg/controllers/management/nodedriver/machine_driver.go [hunk 4]
```diff
@@ -61,7 +61,7 @@ func (m *Lifecycle) download(obj *v3.NodeDriver) (*v3.NodeDriver, error) {
 		return obj, nil
 	}
 
-	_, err = v3.NodeDriverConditionDownloaded.Do(obj, func() (runtime.Object, error) {
+	newObj, err := v3.NodeDriverConditionDownloaded.Once(obj, func() (runtime.Object, error) {
```
Comment (StrongMonkey) [inline pkg/controllers/management/nodedriver/machine_driver.go]: why it is changed from Do to Once?

**Comment by alena1108** [comment]
We need to squash the commits

File: pkg/controllers/management/nodedriver/dynamic_driver.go [hunk 5]
```diff
@@ -176,6 +176,24 @@ func (d *Driver) Exists() bool {
...
+		return nil
+	}
+
+	binaryPath := path.Join(binDir(), d.name)
+	_, err := os.Stat(binaryPath)
+	if err != nil {
+		return fmt.Errorf("Driver %s not found", binaryPath)
+	}
+	err = exec.Command(binaryPath).Start()
+	return errors.Wrapf(err, "Driver binary %s couldn't execute", binaryPath)
```
Comment (alena1108) [inline pkg/controllers/management/nodedriver/dynamic_driver.go]: although perhaps errors.wrapf would return nil when err is nil, this construct doesn't look clean in the code. Lets check for error state, and return nil if error is nil, explicitly

File: pkg/controllers/management/nodedriver/dynamic_driver.go [hunk 6]
```diff
@@ -236,10 +254,20 @@ func (d *Driver) copyBinary(cacheFile, input string) (string, error) {
 		if err != nil {
 			return "", err
 		}
-		driverName = strings.Split(path.Base(u.Path), "_")[0]
-		if !strings.HasPrefix(driverName, "docker-machine-driver-") {
+
+		if !strings.HasPrefix(path.Base(u.Path), "docker-machine-driver-") {
```
Comment (alena1108) [inline pkg/controllers/management/nodedriver/dynamic_driver.go]: lets make a constant for docker-machine-driver as it being referenced multiple times

File: pkg/controllers/management/nodedriver/machine_driver.go (hunk 4 above)
Comment (aiwantaozi) [reply pkg/controllers/management/nodedriver/machine_driver.go]: if may error after download and need retry, when retry, the binary already download, we don't need to do it again, so I change to once.

**Comment by StrongMonkey** [comment]
LGTM

**Pull request merged**